### PR TITLE
Add custom autodiff, energy/gradient compilation, and solver in web-runtime

### PR DESCRIPTION
**Pull request by hypotext**
# Description

Related issues: #213

tf.js was really slow for us because it's designed for huge tensors, small comp graph. We were using it to do operations with a huge comp graph and 1x1 tensors. Not the best use case. So, this PR extends Penrose with a faster solver that compiles the gradient into code from the computation graph, and runs it as code dynamically (the idea is that the JS compiler will JIT it for us, though in practice we haven't confirmed it does). This branch consistently results in a 6x speedup for us over the Haskell system (though is a little more numerically unstable). Also, we can probably do more optimization to get more speedup.

This PR includes:

- [x] Custom reverse-mode autodiff (without Wengert tape implementation)
  - [x] custom differentiable number type, gradient function, and some differentiable ops. All are located in the `Constraints` file
- [x] Compile energy
- [x] Compile gradients
- [x] Compile with hyperparameters
- [x] Integrate compiled energy and gradients into existing solver
- [x] Integrate line search
- [x] Caching results of computation in translation (involving mutable state)
- [x] Port bigger venn example and benchmark
- [x] Port bigger tree example and benchmark
- [x] Port L-BFGS
- [x] Phase out tensorflow.js
- [x] Port continuous map example 
- [x] Port some of geometry example
- [x] Port cross-domain example and benchmark

# Implementation strategy and design decisions

Theory-wise, the code is generated via reverse-mode autodiff, which caches intermediate results. The autodiff builds a mutable computation graph per operation.

High-level documentation for the theory of this work can be found in the Google Doc titled "dynamic gradients" in the Penrose team folder. Other documentation can be found in meeting notes from 6/2020-10/2020.

# Use

This PR doesn't yet include full documentation for how to use the new autodiff system. 

- [ ] Document how to write an example
- [ ] Document the code and working examples
- [ ] Write a high-level document of the new autodiff (incl. use, extension, ops)

# Examples with steps to reproduce them

Working paths:

* Two sets with labels: `runpenrose set-theory-domain/twosets-simple.sub set-theory-domain/venn-small.sty  set-theory-domain/setTheory.dsl`
* ??: `runpenrose set-theory-domain/tree.sub set-theory-domain/venn-small.sty  set-theory-domain/setTheory.dsl`
* ??: `runpenrose set-theory-domain/tree-small2.sub set-theory-domain/venn-opt-test.sty set-theory-domain/setTheory.dsl` 
* Big set example with circles: `runpenrose set-theory-domain/tree.sub set-theory-domain/venn-opt-test.sty set-theory-domain/setTheory.dsl` 
* Big set example with tree:`runpenrose set-theory-domain/tree.sub set-theory-domain/tree.sty set-theory-domain/setTheory.dsl` (except for this:)

image

- [x] Continuous map: `runpenrose set-theory-domain/continuousmap.sub set-theory-domain/continuousmap.sty set-theory-domain/setTheory.dsl`

image

- [x] Hyperbolic: `runpenrose hyperbolic-domain/hyperbolic-example.sub hyperbolic-domain/PoincareDisk.sty hyperbolic-domain/hyperbolic.dsl`

image

- [x] Geometry (partial): `runpenrose geometry-domain/pythagorean-theorem-sugared.sub geometry-domain/euclidean-simple.sty geometry-domain/geometry.dsl`

image

[x] Sets (optimizing thru computation): `runpenrose set-theory-domain/multisets.sub set-theory-domain/venn-comp-test.sty  set-theory-domain/setTheory.dsl`

image

- [x] Cross-domain mesh-set example: `runpenrose mesh-set-domain/DomainInterop.sub mesh-set-domain/DomainInterop.sty mesh-set-domain/DomainInterop.dsl`

For ongoing documentation of performance, see #implementation in Slack.

# Checklist

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran [Haddock](https://github.com/penrose/penrose/wiki/Writing-and-generating-documentation#generating-html-documentation-site) and there were no errors when generating the HTML site
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass locally using `stack test`
- [ ] My code follows the style guidelines of this project (e.g.: no HLint warnings)

# Open questions

Questions that require more discussion or to be addressed in future development:

- [ ] numerical stability of compiled gradients (in particular, use of `EPS_DENOM`) -- do they remain stable on bigger examples? Some examples NaN on sampling
- [ ] use of call stack to compile gradient/energy code -- does it scale to bigger examples, or does it blow the stack?
- [ ] hacks in `clearPacket`
- [ ] make list of default exports in Autodiff
- [ ] separate out tests into jest test suite into CI
- [ ] Benchmark the performance difference between this and C++
- [ ] port shapes to frontend
- [ ] cache generated function on resampling
- [ ] Evaluating the shapes for display is still done via interpretation on VarADs; not compiled
drag handling
- [ ] resample doesn't work w/ new function names unless you add it in the backend
- [ ] analyze and optimize the gradient graph

# Other documentation 

* To check the compiled energies/gradients WRT caching in translation, see these gists: [No caching](https://gist.github.com/hypotext/65fc2b06ec5a9e2fa6fe1e45168343e7), [With caching](https://gist.github.com/hypotext/1012c40bccdb62de64b24cd2ceaafee4)

This PR is a squashed version of this [messy one](https://github.com/penrose/penrose/pull/343).

### CODE REVIEW

**Comment by hypotext** [comment]
Hey @wodeni, is this one-commit branch good to merge into `web-runtime`? (Moving from https://github.com/penrose/penrose/pull/343) If so, feel free to merge it. Otherwise you can approve it and also tell me what to change to comply with the `web-runtime` cleanups :)

File: react-renderer/src/engine/Evaluator.ts [hunk 1]
```diff
@@ -1,9 +1,54 @@
+<<<<<<< HEAD:react-renderer/src/engine/Evaluator.ts
```
Comment (wodeni) [inline react-renderer/src/engine/Evaluator.ts:1]: What happened here????

**Review by wodeni — CHANGES_REQUESTED** [review]

**Comment by hypotext** [comment]
Closing as I forgot to pull from `web-runtime` first, resulting in merge conflicts. Should have built first! Moving to #383.
